zap/src/parser/convert.rs

```diff
@@ -951,86 +1006,133 @@ impl<'src> Converter<'src> {
 		}
 	}
 
-	fn ty_has_unbounded_ref(
+	fn ty_recursion_kind(
 		&self,
-		name: &'src str,
+		target_path: &str,
 		ty: &SyntaxTy<'src>,
-		searched: &mut HashSet<&'src str>,
-	) -> Option<SyntaxIdentifier<'src>> {
+		searched: &mut HashSet<String>,
+	) -> TyRecursionKind<'src> {
 		match &ty.kind {
 			SyntaxTyKind::Arr(ty, len) => {
 				let len = len.map(|len| self.range(&len)).unwrap_or_default();
 
 				// if array does not have a min size of 0, it is unbounded
 				if len.min().is_some_and(|min| min != 0.0) {
-					self.ty_has_unbounded_ref(name, ty, searched)
+					self.ty_recursion_kind(target_path, ty, searched)
 				} else {
-					None
+					TyRecursionKind::None
 				}
 			}
 
 			SyntaxTyKind::Ref(ref_ty) => {
-				let ref_name = ref_ty.name;
+				let key = self
+					.path
+					.iter()
+					.copied()
+					.chain(std::iter::once(ref_ty.name))
+					.collect::<Vec<_>>()
+					.join(".");
 
-				match ref_name {
-					ref_name if ref_name == name => Some(*ref_ty),
+				match &*key {
+					ref_name if ref_name == target_path => TyRecursionKind::Unbounded(*ref_ty),
 
-					"boolean" | "Color3" | "Vector3" | "vector" | "AlignedCFrame" | "CFrame" | "unknown" => None,
+					"boolean" | "Color3" | "Vector3" | "vector" | "AlignedCFrame" | "CFrame" | "unknown" => {
+						TyRecursionKind::None
+					}
```

Comment:
This doesn't cover all the Ref types... which makes me consider if this block is really needed or if we'll get adverse effects with ones that aren't listed here